src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -4792,15 +4797,23 @@ public override BoundNode VisitDeconstructionAssignmentOperator(BoundDeconstruct
             return null;
         }
 
-        private void VisitDeconstructionArguments(ArrayBuilder<DeconstructionVariable> variables, Conversion conversion, BoundExpression right)
+        private void VisitDeconstructionArguments(ArrayBuilder<DeconstructionVariable> variables, Conversion conversion, BoundExpression right, TypeWithState? rightResultOpt = null)
         {
             Debug.Assert(conversion.Kind == ConversionKind.Deconstruction);
 
             int n = variables.Count;
 
             if (!conversion.DeconstructionInfo.IsDefault)
             {
-                VisitRvalue(right);
+                // If we were passed an explicit right result, use that instead of visiting to figure it out
+                if (rightResultOpt.HasValue)
+                {
+                    SetResultType(right, rightResultOpt.Value);
+                }
+                else
+                {
+                    VisitRvalue(right);
```

Comment:
>VisitRvalue [](start = 20, length = 11)

We need to do this unconditionally, and then update the result type if necessary, so that we can record information for any nested expressions in `right`.

Reply:
Makes sense.